### PR TITLE
Refactor multistream support

### PR DESCRIPTION
**Pull request by benson31**
This puts all communications on unique-to-backend streams. I need to test this with LBANN yet, due to changes in the Aluminum communicator wrapper.

This obviates the need for more than 2 additional compute streams.

Future work/optimizations after this PR:
- TT algorithms
- Use the input matrix stream(s) before resorting to the stream pool
- Adjust algorithms to never use more than 2 streams rather than the whole pool (I'm willing to do this sooner if the reviewers think it's best done now. I have not noticed a performance degradation using >2 streams, but also no benefit.)

### CODE REVIEW

File: include/El/core/imports/aluminum.hpp [hunk 1]
```diff
@@ -5,10 +5,18 @@
...
+#include "nvToolsExt.h"
+#include "nvToolsExtCuda.h"
+#include "nvToolsExtCudaRt.h"
+#endif // HYDROGEN_HAVE_NVPROF
+
 #endif // HYDROGEN_HAVE_ALUMINUM
 
 namespace El
 {
+void break_on_me();
```
Comment (timmoon10) [inline include/El/core/imports/aluminum.hpp]: Where is the implementation? Also, we should move it inside `El::mpi::internal`.

File: src/core/imports/mpi/AllToAll.hpp.save [hunk 2]
```diff
@@ -0,0 +1,239 @@
+// AllToAll
```
Comment (timmoon10) [inline src/core/imports/mpi/AllToAll.hpp.save]: Is this file supposed to be included?

**Review by timmoon10 — APPROVED** [review]
Looks sensible to me. For SUMMA-C, was there any difference when A and B do their communication on different streams? Does having more "stream teams" make up for the loss of overlap?

File: include/El/core/imports/aluminum.hpp (hunk 1 above)
Comment (benson31) [reply include/El/core/imports/aluminum.hpp]: It's a function in `include/El/core/environment/impl.hpp`. I just need the decl here. I could move it up in the world, but that's more effort than, IMO, it's worth at this point or in this PR...

File: src/core/imports/mpi/AllToAll.hpp.save (hunk 2 above)
Comment (benson31) [reply src/core/imports/mpi/AllToAll.hpp.save]: No. Removed.

**Comment by benson31** [comment]
LBANN builds and runs.

File: include/El/core/imports/aluminum.hpp [hunk 3]
```diff
@@ -5,10 +5,19 @@
...
+#include "nvToolsExtCuda.h"
+#include "nvToolsExtCudaRt.h"
+#endif // HYDROGEN_HAVE_NVPROF
+
 #endif // HYDROGEN_HAVE_ALUMINUM
 
 namespace El
 {
+// "Real" declaration is in include/El/core/environment/impl.hpp
+extern void break_on_me();
```
Comment (bvanessen) [inline include/El/core/imports/aluminum.hpp:20]: Is this supposed to be there, or part of debugging code?

File: include/El/core/imports/mpi/error.hpp [hunk 4]
```diff
@@ -111,7 +111,7 @@ inline void MpiError(
     int mpi_error_code, std::string const& function,
     std::string const& file, int line)
 {
-    //El::break_on_me();
+    El::break_on_me();
```
Comment (bvanessen) [inline include/El/core/imports/mpi/error.hpp:114]: Again, seems like this should be commented out.

**Review by bvanessen — APPROVED** [review]
LGTM

File: include/El/core/imports/aluminum.hpp (hunk 3 above)
Comment (benson31) [reply include/El/core/imports/aluminum.hpp:20]: It's supposed to be there.

File: include/El/core/imports/mpi/error.hpp (hunk 4 above)
Comment (benson31) [reply include/El/core/imports/mpi/error.hpp:114]: It's supposed to be there.

**Pull request merged**